src/lem/circuit.rs

```diff
@@ -701,6 +705,100 @@ fn synthesize_block<F: LurkField, CS: ConstraintSystem<F>, C: Coprocessor<F>>(
             Op::Decons4(preimg, img) => {
                 decons_helper!(preimg, img, SlotType::Hash8);
             }
+            Op::PushBinding(img, preimg) => {
+                // Retrieve allocated preimage
+                let sym = bound_allocations.get_ptr(&preimg[0])?;
+                let val = bound_allocations.get_ptr(&preimg[1])?;
+                let env = bound_allocations.get_ptr(&preimg[2])?;
+
+                // Retrieve the preallocated preimage and image for this slot
+                let (preallocated_preimg, preallocated_img_hash) =
+                    &ctx.hash4_slots[next_slot.consume_hash4()];
+
+                // For each component of the preimage, add implication constraints
+                // for its tag and hash
+                let sym_tag = ctx.global_allocator.alloc_tag_cloned(&mut cs, &Sym);
+                let env_tag = ctx.global_allocator.alloc_tag_cloned(&mut cs, &Env);
+                implies_equal(
+                    &mut cs.namespace(|| format!("implies equal {}.tag", preimg[0])),
+                    not_dummy,
+                    sym.tag(),
+                    &sym_tag,
+                );
+                implies_equal(
+                    &mut cs.namespace(|| format!("implies equal {}.hash", preimg[0])),
+                    not_dummy,
+                    sym.hash(),
+                    &preallocated_preimg[0],
+                );
+                implies_equal(
+                    &mut cs.namespace(|| format!("implies equal {}.tag", preimg[1])),
+                    not_dummy,
+                    val.tag(),
+                    &preallocated_preimg[1],
+                );
+                implies_equal(
+                    &mut cs.namespace(|| format!("implies equal {}.hash", preimg[1])),
+                    not_dummy,
+                    val.hash(),
+                    &preallocated_preimg[2],
+                );
+                implies_equal(
+                    &mut cs.namespace(|| format!("implies equal {}.tag", preimg[2])),
+                    not_dummy,
+                    env.tag(),
+                    &env_tag,
+                );
+                implies_equal(
+                    &mut cs.namespace(|| format!("implies equal {}.hash", preimg[2])),
+                    not_dummy,
+                    env.hash(),
+                    &preallocated_preimg[3],
+                );
+
+                // Allocate the image tag if it hasn't been allocated before,
+                // create the full image pointer and add it to bound allocations
+                let img_tag = ctx.global_allocator.alloc_tag_cloned(&mut cs, &Env);
```

Comment:
Is this 'maybe' allocation deterministic within the circuit, or does it depend on the dynamic data? If the latter, then we risk inconsistent circuits. For this to be okay, the calls to `alloc_tag_cloned` need to always happen in the same order, with the same arguments. This is a larger question than just this usage.

I assume it's okay, since the design of the global allocator (and these lazy-allocating methods) assumes the above. I'm just mentioning so you can tell me it's fine everywhere. If in any doubt, we should convince ourselves.

Reply:
I believe that all constants, including tags, are allocated before, so when you call this, it only amounts to getting the previously allocated constant. This does not depend on dynamic data, so it's fine